File: ios/Classes/GalaceanPlayerView.swift
import Flutter
import UIKit
import GalaceanEffects

/**
 * Galacean Player View
 * iOS 端的播放器视图实现，集成 Galacean Effects Native SDK
 *
 * SDK 类：GEPlayer, GEPlayerParams
 */
class GalaceanPlayerView: NSObject, FlutterPlatformView {
    private var containerView: UIView
    private var methodChannel: FlutterMethodChannel
    
    // Galacean Player 实例
    private var gePlayer: GEPlayer?
    private var playerParams: GEPlayerParams?
    
    private var isPlaying = false
    private var isLooping = false
    private var repeatCount: Int32 = 0  // 0 表示无限循环，>0 表示播放次数
    private var playbackSpeed: Double = 1.0
    private var currentSceneUrl: String?
    private var isSceneLoaded = false
    
    init(
        frame: CGRect,
        viewId: Int64,
        messenger: FlutterBinaryMessenger,
        args: Any?
    ) {
        containerView = UIView(frame: frame)
        containerView.backgroundColor = .clear
        containerView.clipsToBounds = true
        
        // 创建方法通道
        methodChannel = FlutterMethodChannel(
            name: "galacean_native_player_\(viewId)",
            binaryMessenger: messenger
        )
        
        super.init()
        
        // 设置方法调用处理
        methodChannel.setMethodCallHandler(handle)
        
        print("GalaceanPlayerView created, id: \(viewId)")
    }
    
    func view() -> UIView {
        return containerView
    }
    
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadScene":
            guard let args = call.arguments as? [String: Any],
                  let url = args["url"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT",
                                  message: "URL cannot be null",
                                  details: nil))
                return
            }
            let autoPlay = args["autoPlay"] as? Bool ?? true
            loadScene(url: url, autoPlay: autoPlay, result: result)
            
        case "play":
            play(result: result)
            
        case "pause":
            pause(result: result)
            
        case "resume":
            resume(result: result)
            
        case "stop":
            stop(result: result)
            
        case "replay":
            replay(result: result)
            
        case "setLoop":
            guard let args = call.arguments as? [String: Any],
                  let loop = args["loop"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGUMENT",
                                  message: "Loop parameter is required",
                                  details: nil))
                return
            }
            setLoop(loop: loop, result: result)
            
        case "setSpeed":
            guard let args = call.arguments as? [String: Any],
                  let speed = args["speed"] as? Double else {
                result(FlutterError(code: "INVALID_ARGUMENT",
                                  message: "Speed parameter is required",
                                  details: nil))
                return
            }
            setSpeed(speed: speed, result: result)
            
        case "getCurrentTime":
            getCurrentTime(result: result)
            
        case "getDuration":
            getDuration(result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Player Control Methods
    
    /**
     * 加载场景
     */
    private func loadScene(url: String, autoPlay: Bool, result: @escaping FlutterResult) {
        currentSceneUrl = url
        print("Loading scene: \(url), autoPlay: \(autoPlay)")
        
        invokeFlutterMethod("onStateChanged", arguments: "loading")
        
        // 销毁旧的播放器
        destroyPlayer()
        
        // 创建播放参数
        playerParams = GEPlayerParams()
        playerParams?.url = url
        
        // 创建 GEPlayer
        gePlayer = GEPlayer(params: playerParams!)
        
        // 设置 frame 并添加到容器
        gePlayer?.frame = containerView.bounds
        gePlayer?.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        containerView.addSubview(gePlayer!)
        
        // 加载场景
        gePlayer?.loadScene { [weak self] success, errorMsg in
            guard let self = self else { return }
            
            DispatchQueue.main.async {
                if success {
                    print("Scene loaded successfully")
                    self.isSceneLoaded = true
                    self.invokeFlutterMethod("onLoadComplete", arguments: nil)
                    self.invokeFlutterMethod("onStateChanged", arguments: "ready")
                    
                    if autoPlay {
                        self.playInternal(result: nil)
                    }
                    result(nil)
                } else {
                    print("Failed to load scene: \(errorMsg ?? "Unknown error")")
                    self.isSceneLoaded = false
                    let error = errorMsg ?? "Unknown error"
                    self.invokeFlutterMethod("onError", arguments: error)
                    result(FlutterError(code: "LOAD_FAILED",
                                      message: error,
                                      details: nil))
                }
            }
        }
    }
    
    /**
     * 播放
     */
    private func play(result: FlutterResult?) {
        playInternal(result: result)
    }
    
    /**
     * 内部播放方法
     */
    private func playInternal(result: FlutterResult?) {
        guard let player = gePlayer else {
            result?(FlutterError(code: "NOT_INITIALIZED",
                               message: "Player not initialized",
                               details: nil))
            return
        }
        
        guard isSceneLoaded else {
            result?(FlutterError(code: "SCENE_NOT_LOADED",
                               message: "Scene not loaded yet",
                               details: nil))
            return
        }
        
        print("Play, repeatCount: \(repeatCount)")
        
        player.play(withRepeatCount: repeatCount) { [weak self] success, errorMsg in
            guard let self = self else { return }
            
            DispatchQueue.main.async {
                if success {
                    print("Play completed")
                    self.isPlaying = false
                    self.invokeFlutterMethod("onStateChanged", arguments: "stopped")
                    self.invokeFlutterMethod("onPlayComplete", arguments: nil)
                } else {
                    print("Play error: \(errorMsg ?? "Unknown error")")
                    self.isPlaying = false
                    self.invokeFlutterMethod("onError", arguments: errorMsg ?? "Play failed")
                }
            }
        }
        
        isPlaying = true
        invokeFlutterMethod("onStateChanged", arguments: "playing")
        result?(nil)
    }
    
    /**
     * 暂停
     */
    private func pause(result: @escaping FlutterResult) {
        guard let player = gePlayer else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Player not initialized",
                              details: nil))
            return
        }
        
        print("Pause")
        player.pause()
        isPlaying = false
        invokeFlutterMethod("onStateChanged", arguments: "paused")
        result(nil)
    }
    
    /**
     * 恢复播放
     */
    private func resume(result: @escaping FlutterResult) {
        guard let player = gePlayer else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Player not initialized",
                              details: nil))
            return
        }
        
        print("Resume")
        player.resume()
        isPlaying = true
        invokeFlutterMethod("onStateChanged", arguments: "playing")
        result(nil)
    }
    
    /**
     * 停止
     */
    private func stop(result: @escaping FlutterResult) {
        guard let player = gePlayer else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Player not initialized",
                              details: nil))
            return
        }
        
        print("Stop")
        player.stop()
        isPlaying = false
        invokeFlutterMethod("onStateChanged", arguments: "stopped")
        result(nil)
    }
    
    /**
     * 重新播放
     */
    private func replay(result: @escaping FlutterResult) {
        guard let player = gePlayer else {
            result(FlutterError(code: "NOT_INITIALIZED",
                              message: "Player not initialized",
                              details: nil))
            return
        }
        
        print("Replay")
        player.stop()
        playInternal(result: result)
    }
    
    /**
     * 设置循环播放
     */
    private func setLoop(loop: Bool, result: @escaping FlutterResult) {
        print("Set loop: \(loop)")
        isLooping = loop
        // 0 表示无限循环，1 表示播放一次
        repeatCount = loop ? 0 : 1
        result(nil)
    }
    
    /**
     * 设置播放速度
     * 注意：当前 SDK 可能不支持此功能
     */
    private func setSpeed(speed: Double, result: @escaping FlutterResult) {
        print("Set speed: \(speed) (not supported by SDK)")
        playbackSpeed = speed
        // SDK 可能不支持设置播放速度
        result(nil)
    }
    
    /**
     * 获取当前播放时间
     * 注意：SDK 没有提供此方法，返回 0
     */
    private func getCurrentTime(result: @escaping FlutterResult) {
        // SDK 没有提供获取当前时间的方法
        let time = 0.0
        result(time)
    }
    
    /**
     * 获取总时长
     * 可以通过 frameCount 估算
     */
    private func getDuration(result: @escaping FlutterResult) {
        guard let player = gePlayer else {
            result(0.0)
            return
        }
        
        let frameCount = player.getFrameCount()
        // 假设 30fps，计算时长（秒）
        let duration = Double(frameCount) / 30.0
        result(duration)
    }
    
    // MARK: - Helper Methods
    
    /**
     * 销毁播放器
     */
    private func destroyPlayer() {
        gePlayer?.destroy()
        gePlayer?.removeFromSuperview()
        gePlayer = nil
        playerParams = nil
        isSceneLoaded = false
        isPlaying = false
    }
    
    /**
     * 辅助方法：调用 Flutter 方法
     */
    private func invokeFlutterMethod(_ method: String, arguments: Any?) {
        DispatchQueue.main.async { [weak self] in
            self?.methodChannel.invokeMethod(method, arguments: arguments)
        }
    }
    
    // MARK: - Cleanup
    
    deinit {
        print("Disposing player")
        destroyPlayer()
    }
}

File: android/src/main/kotlin/github/kaierwen/galacean/galacean_native_player/GalaceanPlayerView.kt
package github.kaierwen.galacean.galacean_native_player

import android.content.Context
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import com.antgroup.galacean.effects.GEPlayer
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * Galacean Player View
 * Android 端的播放器视图实现，集成 Galacean Effects Native SDK
 * 
 * SDK 包名: com.antgroup.galacean.effects
 * 主要类: GEPlayer, GEPlayer.GEPlayerParams, GEPlayer.Callback
 */
class GalaceanPlayerView(
    private val context: Context,
    private val id: Int,
    messenger: BinaryMessenger
) : PlatformView, MethodChannel.MethodCallHandler {

    companion object {
        private const val TAG = "GalaceanPlayerView"
    }

    private val container: FrameLayout = FrameLayout(context)
    private val methodChannel: MethodChannel
    private val mainHandler = Handler(Looper.getMainLooper())
    
    // Galacean Player 实例
    private var gePlayer: GEPlayer? = null
    
    private var isPlaying = false
    private var isLooping = false
    private var repeatCount = 0  // 0 表示无限循环，>0 表示播放次数
    private var playbackSpeed = 1.0
    private var currentSceneUrl: String? = null
    private var isInitialized = false
    private var isSceneLoaded = false

    init {
        // 创建方法通道
        methodChannel = MethodChannel(messenger, "galacean_native_player_$id")
        methodChannel.setMethodCallHandler(this)
        
        // 设置容器背景色
        container.setBackgroundColor(Color.TRANSPARENT)
        
        isInitialized = true
        Log.d(TAG, "GalaceanPlayerView created, id: $id")
    }

    override fun getView(): View {
        return container
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadScene" -> {
                val url = call.argument<String>("url")
                val autoPlay = call.argument<Boolean>("autoPlay") ?: true
                if (url != null) {
                    loadScene(url, autoPlay, result)
                } else {
                    result.error("INVALID_ARGUMENT", "URL cannot be null", null)
                }
            }
            "play" -> {
                play(result)
            }
            "pause" -> {
                pause(result)
            }
            "resume" -> {
                resume(result)
            }
            "stop" -> {
                stop(result)
            }
            "replay" -> {
                replay(result)
            }
            "setLoop" -> {
                val loop = call.argument<Boolean>("loop") ?: false
                setLoop(loop, result)
            }
            "setSpeed" -> {
                val speed = call.argument<Double>("speed") ?: 1.0
                setSpeed(speed, result)
            }
            "getCurrentTime" -> {
                getCurrentTime(result)
            }
            "getDuration" -> {
                getDuration(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    /**
     * 加载场景
     */
    private fun loadScene(url: String, autoPlay: Boolean, result: MethodChannel.Result) {
        if (!isInitialized) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            currentSceneUrl = url
            Log.d(TAG, "Loading scene: $url, autoPlay: $autoPlay")
            
            invokeFlutterMethod("onStateChanged", "loading")
            
            // 先销毁旧的播放器
            destroyPlayer()
            
            // 创建播放参数
            val params = GEPlayer.GEPlayerParams().apply {
                this.url = url
            }
            
            // 创建新的 GEPlayer
            gePlayer = GEPlayer(context, params)
            
            // 添加到容器
            container.removeAllViews()
            container.addView(gePlayer, FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            ))
            
            // 加载场景
            gePlayer?.loadScene(object : GEPlayer.Callback {
                override fun onResult(success: Boolean, errorMsg: String?) {
                    mainHandler.post {
                        if (success) {
                            Log.d(TAG, "Scene loaded successfully")
                            isSceneLoaded = true
                            invokeFlutterMethod("onLoadComplete", null)
                            invokeFlutterMethod("onStateChanged", "ready")
                            
                            if (autoPlay) {
                                playInternal(null)
                            }
                            result.success(null)
                        } else {
                            Log.e(TAG, "Failed to load scene: $errorMsg")
                            isSceneLoaded = false
                            val error = errorMsg ?: "Unknown error"
                            invokeFlutterMethod("onError", error)
                            result.error("LOAD_FAILED", error, null)
                        }
                    }
                }
            })
            
        } catch (e: Exception) {
            Log.e(TAG, "Load scene exception", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result.error("LOAD_ERROR", error, null)
        }
    }

    /**
     * 播放
     */
    private fun play(result: MethodChannel.Result?) {
        playInternal(result)
    }
    
    /**
     * 内部播放方法
     */
    private fun playInternal(result: MethodChannel.Result?) {
        if (!isInitialized || gePlayer == null) {
            result?.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        if (!isSceneLoaded) {
            result?.error("SCENE_NOT_LOADED", "Scene not loaded yet", null)
            return
        }
        
        try {
            Log.d(TAG, "Play, repeatCount: $repeatCount")
            
            gePlayer?.play(repeatCount, object : GEPlayer.Callback {
                override fun onResult(success: Boolean, errorMsg: String?) {
                    mainHandler.post {
                        if (success) {
                            Log.d(TAG, "Play completed")
                            isPlaying = false
                            invokeFlutterMethod("onStateChanged", "stopped")
                            invokeFlutterMethod("onPlayComplete", null)
                        } else {
                            Log.e(TAG, "Play error: $errorMsg")
                            isPlaying = false
                            invokeFlutterMethod("onError", errorMsg ?: "Play failed")
                        }
                    }
                }
            })
            
            isPlaying = true
            invokeFlutterMethod("onStateChanged", "playing")
            result?.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Play error", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result?.error("PLAY_ERROR", error, null)
        }
    }

    /**
     * 暂停
     */
    private fun pause(result: MethodChannel.Result) {
        if (!isInitialized || gePlayer == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            Log.d(TAG, "Pause")
            gePlayer?.pause()
            isPlaying = false
            invokeFlutterMethod("onStateChanged", "paused")
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Pause error", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result.error("PAUSE_ERROR", error, null)
        }
    }

    /**
     * 恢复播放
     */
    private fun resume(result: MethodChannel.Result) {
        if (!isInitialized || gePlayer == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            Log.d(TAG, "Resume")
            gePlayer?.resume()
            isPlaying = true
            invokeFlutterMethod("onStateChanged", "playing")
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Resume error", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result.error("RESUME_ERROR", error, null)
        }
    }

    /**
     * 停止
     */
    private fun stop(result: MethodChannel.Result) {
        if (!isInitialized || gePlayer == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            Log.d(TAG, "Stop")
            gePlayer?.stop()
            isPlaying = false
            invokeFlutterMethod("onStateChanged", "stopped")
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Stop error", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result.error("STOP_ERROR", error, null)
        }
    }

    /**
     * 重新播放
     */
    private fun replay(result: MethodChannel.Result) {
        if (!isInitialized || gePlayer == null) {
            result.error("NOT_INITIALIZED", "Player not initialized", null)
            return
        }
        
        try {
            Log.d(TAG, "Replay")
            gePlayer?.stop()
            playInternal(result)
        } catch (e: Exception) {
            Log.e(TAG, "Replay error", e)
            val error = e.message ?: "Unknown error"
            invokeFlutterMethod("onError", error)
            result.error("REPLAY_ERROR", error, null)
        }
    }

    /**
     * 设置循环播放
     */
    private fun setLoop(loop: Boolean, result: MethodChannel.Result) {
        try {
            Log.d(TAG, "Set loop: $loop")
            isLooping = loop
            // 0 表示无限循环，1 表示播放一次
            repeatCount = if (loop) 0 else 1
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Set loop error", e)
            val error = e.message ?: "Unknown error"
            result.error("SET_LOOP_ERROR", error, null)
        }
    }

    /**
     * 设置播放速度
     * 注意：当前 SDK 可能不支持此功能
     */
    private fun setSpeed(speed: Double, result: MethodChannel.Result) {
        try {
            Log.d(TAG, "Set speed: $speed (not supported by SDK)")
            playbackSpeed = speed
            // SDK 可能不支持设置播放速度
            result.success(null)
        } catch (e: Exception) {
            Log.e(TAG, "Set speed error", e)
            val error = e.message ?: "Unknown error"
            result.error("SET_SPEED_ERROR", error, null)
        }
    }

    /**
     * 获取当前播放时间
     * 注意：SDK 没有提供此方法，返回 0
     */
    private fun getCurrentTime(result: MethodChannel.Result) {
        try {
            // SDK 没有提供获取当前时间的方法
            val time = 0.0
            result.success(time)
        } catch (e: Exception) {
            Log.e(TAG, "Get current time error", e)
            result.error("GET_TIME_ERROR", e.message, null)
        }
    }

    /**
     * 获取总时长
     * 可以通过 frameCount 估算
     */
    private fun getDuration(result: MethodChannel.Result) {
        try {
            val frameCount = gePlayer?.frameCount ?: 0
            // 假设 30fps，计算时长（秒）
            val duration = frameCount / 30.0
            result.success(duration)
        } catch (e: Exception) {
            Log.e(TAG, "Get duration error", e)
            result.error("GET_DURATION_ERROR", e.message, null)
        }
    }
    
    /**
     * 销毁播放器
     */
    private fun destroyPlayer() {
        try {
            gePlayer?.destroy()
            gePlayer = null
            isSceneLoaded = false
            isPlaying = false
        } catch (e: Exception) {
            Log.e(TAG, "Error destroying player", e)
        }
    }
    
    /**
     * 辅助方法：在主线程调用 Flutter 方法
     */
    private fun invokeFlutterMethod(method: String, arguments: Any?) {
        mainHandler.post {
            try {
                methodChannel.invokeMethod(method, arguments)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to invoke Flutter method: $method", e)
            }
        }
    }

    override fun dispose() {
        Log.d(TAG, "Disposing player")
        destroyPlayer()
        container.removeAllViews()
        methodChannel.setMethodCallHandler(null)
    }
}
